File: ios/Earwyrm/Features/Explore/ExploreViewModel.swift
import Foundation
import Supabase

// MARK: - Filter Types

enum TimeRange: String, CaseIterable {
    case all = "all time"
    case week = "this week"
    case today = "today"

    var cutoffDate: Date? {
        switch self {
        case .all: return nil
        case .week: return Calendar.current.date(byAdding: .day, value: -7, to: Date())
        case .today: return Calendar.current.date(byAdding: .day, value: -1, to: Date())
        }
    }
}

enum SortOption: String, CaseIterable {
    case newest = "Newest"
    case resonated = "Most resonated"
    case discussed = "Most discussed"
}

// MARK: - Trending Tag

struct TrendingTag: Identifiable {
    let tag: String
    let count: Int
    var id: String { tag }
}

// MARK: - ViewModel

@Observable
@MainActor
final class ExploreViewModel {
    var allLyrics: [Lyric] = []
    var profileMap: [UUID: String] = [:]
    var trendingTags: [TrendingTag] = []
    var isLoading = false
    var error: String?

    private static let lyricColumns = """
        id, user_id, content, song_title, artist_name, cover_art_url, \
        album_name, is_current, is_public, tags, share_token, \
        canonical_lyric_id, musicbrainz_recording_id, musicbrainz_release_id, \
        reaction_count, comment_count, created_at, replaced_at
        """

    // MARK: - Fetching

    func fetchPublicLyrics() async {
        isLoading = true
        error = nil
        do {
            let lyrics: [Lyric] = try await supabase
                .from("lyrics")
                .select(Self.lyricColumns)
                .eq("is_public", value: true)
                .order("created_at", ascending: false)
                .limit(80)
                .execute()
                .value
            allLyrics = lyrics

            // Batch-fetch profiles for usernames
            let userIds = Array(Set(lyrics.map(\.userId)))
            if !userIds.isEmpty {
                let profiles: [CommentProfile] = try await supabase
                    .from("profiles")
                    .select("id, username")
                    .in("id", values: userIds.map(\.uuidString))
                    .execute()
                    .value
                for p in profiles {
                    profileMap[p.id] = p.username
                }
            }
        } catch {
            self.error = error.localizedDescription
            print("Fetch public lyrics error: \(error)")
        }
        isLoading = false
    }

    func fetchTrendingTags() async {
        do {
            let lyrics: [Lyric] = try await supabase
                .from("lyrics")
                .select("id, tags, created_at")
                .eq("is_public", value: true)
                .order("created_at", ascending: false)
                .limit(200)
                .execute()
                .value

            var tagCounts: [String: Int] = [:]
            for lyric in lyrics {
                for tag in lyric.tags ?? [] {
                    tagCounts[tag, default: 0] += 1
                }
            }

            trendingTags = tagCounts
                .sorted { $0.value > $1.value }
                .prefix(10)
                .map { TrendingTag(tag: $0.key, count: $0.value) }
        } catch {
            print("Fetch trending tags error: \(error)")
        }
    }

    // MARK: - ForYou Pipeline

    func forYouFeed(
        follows: [Follow],
        userId: UUID?,
        selectedTags: Set<String>,
        timeRange: TimeRange,
        search: String,
        sort: SortOption,
        limit: Int
    ) -> [Lyric] {
        var lyrics = allLyrics

        // Exclude own lyrics
        if let userId {
            lyrics = lyrics.filter { $0.userId != userId }
        }

        // Exclude followed content (ForYou = discovery)
        if !follows.isEmpty {
            let followedArtists = Set(
                follows.filter { $0.filterType == "artist" }
                    .map { $0.filterValue.lowercased() }
            )
            let followedSongs = Set(
                follows.filter { $0.filterType == "song" }
                    .map { $0.filterValue.lowercased() }
            )
            let followedTags = Set(
                follows.filter { $0.filterType == "tag" }
                    .map { $0.filterValue.lowercased() }
            )

            lyrics = lyrics.filter { lyric in
                if lyric.artistName.map({ followedArtists.contains($0.lowercased()) }) ?? false {
                    return false
                }
                if lyric.songTitle.map({ followedSongs.contains($0.lowercased()) }) ?? false {
                    return false
                }
                if lyric.tags?.contains(where: { followedTags.contains($0.lowercased()) }) ?? false {
                    return false
                }
                return true
            }
        }

        // Tag filter
        if !selectedTags.isEmpty {
            lyrics = lyrics.filter { lyric in
                guard let tags = lyric.tags else { return false }
                return tags.contains(where: { selectedTags.contains($0) })
            }
        }

        // Time range
        if let cutoff = timeRange.cutoffDate {
            lyrics = lyrics.filter { $0.createdAt >= cutoff }
        }

        // Search
        lyrics = applySearch(lyrics, query: search)

        // Sort
        lyrics = applySort(lyrics, sort: sort)

        return Array(lyrics.prefix(limit))
    }

    // MARK: - Following Pipeline

    func followingFeed(
        follows: [Follow],
        activeFollowIds: Set<UUID>,
        timeRange: TimeRange,
        search: String,
        sort: SortOption,
        limit: Int,
        userId: UUID?
    ) -> [Lyric] {
        var lyrics = allLyrics

        // Exclude own
        if let userId {
            lyrics = lyrics.filter { $0.userId != userId }
        }

        // Match against follows (all if none active, otherwise only active)
        let followsToMatch = activeFollowIds.isEmpty
            ? follows
            : follows.filter { activeFollowIds.contains($0.id) }

        lyrics = lyrics.filter { lyric in
            matchesAnyFollow(lyric: lyric, follows: followsToMatch)
        }

        // Time range
        if let cutoff = timeRange.cutoffDate {
            lyrics = lyrics.filter { $0.createdAt >= cutoff }
        }

        // Search
        lyrics = applySearch(lyrics, query: search)

        // Sort
        lyrics = applySort(lyrics, sort: sort)

        return Array(lyrics.prefix(limit))
    }

    // MARK: - Helpers

    private func matchesAnyFollow(lyric: Lyric, follows: [Follow]) -> Bool {
        for follow in follows {
            switch follow.filterType {
            case "artist":
                if let artist = lyric.artistName,
                   artist.caseInsensitiveCompare(follow.filterValue) == .orderedSame {
                    return true
                }
            case "song":
                if let song = lyric.songTitle,
                   song.caseInsensitiveCompare(follow.filterValue) == .orderedSame {
                    return true
                }
            case "tag":
                if let tags = lyric.tags,
                   tags.contains(where: { $0.caseInsensitiveCompare(follow.filterValue) == .orderedSame }) {
                    return true
                }
            default:
                break
            }
        }
        return false
    }

    private func applySearch(_ lyrics: [Lyric], query: String) -> [Lyric] {
        guard !query.isEmpty else { return lyrics }
        let q = query.lowercased()
        return lyrics.filter { lyric in
            lyric.content.lowercased().contains(q)
            || (lyric.artistName?.lowercased().contains(q) ?? false)
            || (lyric.songTitle?.lowercased().contains(q) ?? false)
            || (lyric.tags?.contains(where: { $0.lowercased().contains(q) }) ?? false)
        }
    }

    private func applySort(_ lyrics: [Lyric], sort: SortOption) -> [Lyric] {
        switch sort {
        case .newest:
            return lyrics.sorted { $0.createdAt > $1.createdAt }
        case .resonated:
            return lyrics.sorted { ($0.reactionCount ?? 0) > ($1.reactionCount ?? 0) }
        case .discussed:
            return lyrics.sorted { ($0.commentCount ?? 0) > ($1.commentCount ?? 0) }
        }
    }
}

File: android/app/src/main/java/com/earwyrm/app/feature/explore/ExploreViewModel.kt
package com.earwyrm.app.feature.explore

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.earwyrm.app.core.auth.AuthManager
import com.earwyrm.app.core.model.Lyric
import com.earwyrm.app.core.model.Profile
import com.earwyrm.app.core.supabase.BlockManager
import com.earwyrm.app.core.supabase.FollowManager
import com.earwyrm.app.core.model.Reaction
import com.earwyrm.app.core.model.ReactionInsert
import dagger.hilt.android.lifecycle.HiltViewModel
import io.github.jan.supabase.SupabaseClient
import io.github.jan.supabase.postgrest.postgrest
import io.github.jan.supabase.postgrest.query.Order
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.launch
import kotlinx.datetime.Clock
import javax.inject.Inject
import kotlin.time.Duration.Companion.days
import kotlin.time.Duration.Companion.hours

enum class SortOption(val label: String) {
    NEWEST("Newest"),
    MOST_RESONATED("Most Resonated"),
    MOST_DISCUSSED("Most Discussed")
}

enum class TimeRange(val label: String) {
    ALL_TIME("All Time"),
    THIS_MONTH("This Month"),
    THIS_WEEK("This Week"),
    TODAY("Today")
}

data class SearchResults(
    val songs: List<Pair<String, String?>> = emptyList(), // title to artist
    val artists: List<String> = emptyList(),
    val users: List<Profile> = emptyList(),
    val lyrics: List<Lyric> = emptyList()
) {
    val isEmpty: Boolean get() = songs.isEmpty() && artists.isEmpty() && users.isEmpty() && lyrics.isEmpty()
}

@HiltViewModel
class ExploreViewModel @Inject constructor(
    private val supabase: SupabaseClient,
    private val authManager: AuthManager,
    private val followManager: FollowManager,
    private val blockManager: BlockManager
) : ViewModel() {

    private val _forYouLyrics = MutableStateFlow<List<Lyric>>(emptyList())
    val forYouLyrics: StateFlow<List<Lyric>> = _forYouLyrics.asStateFlow()

    private val _followingLyrics = MutableStateFlow<List<Lyric>>(emptyList())
    val followingLyrics: StateFlow<List<Lyric>> = _followingLyrics.asStateFlow()

    private val _lyricProfiles = MutableStateFlow<Map<String, Profile>>(emptyMap())
    val lyricProfiles: StateFlow<Map<String, Profile>> = _lyricProfiles.asStateFlow()

    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()

    private val _selectedTab = MutableStateFlow(0)
    val selectedTab: StateFlow<Int> = _selectedTab.asStateFlow()

    // Search & filter state
    private val _searchQuery = MutableStateFlow("")
    val searchQuery: StateFlow<String> = _searchQuery.asStateFlow()

    private val _sortOption = MutableStateFlow(SortOption.NEWEST)
    val sortOption: StateFlow<SortOption> = _sortOption.asStateFlow()

    private val _timeRange = MutableStateFlow(TimeRange.ALL_TIME)
    val timeRange: StateFlow<TimeRange> = _timeRange.asStateFlow()

    private val _trendingTags = MutableStateFlow<List<Pair<String, Int>>>(emptyList())
    val trendingTags: StateFlow<List<Pair<String, Int>>> = _trendingTags.asStateFlow()

    private val _selectedTag = MutableStateFlow<String?>(null)
    val selectedTag: StateFlow<String?> = _selectedTag.asStateFlow()

    // Search results
    private val _searchResults = MutableStateFlow(SearchResults())
    val searchResults: StateFlow<SearchResults> = _searchResults.asStateFlow()

    private val _isSearching = MutableStateFlow(false)
    val isSearching: StateFlow<Boolean> = _isSearching.asStateFlow()

    val isSearchActive: StateFlow<Boolean> = _searchQuery.map { query ->
        query.isNotBlank()
    }.stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), false)

    // Reaction state: set of lyric IDs the current user has reacted to
    private val _reactedLyricIds = MutableStateFlow<Set<String>>(emptySet())
    val reactedLyricIds: StateFlow<Set<String>> = _reactedLyricIds.asStateFlow()

    // Local reaction count overrides (lyricId -> delta from server count)
    private val _reactionCountDeltas = MutableStateFlow<Map<String, Int>>(emptyMap())
    val reactionCountDeltas: StateFlow<Map<String, Int>> = _reactionCountDeltas.asStateFlow()

    // Filtered versions (only apply sort/time/tag, not search query — search has its own results)
    val filteredForYouLyrics: StateFlow<List<Lyric>> = combine(
        _forYouLyrics, _sortOption, _timeRange, _selectedTag
    ) { lyrics, sort, time, tag ->
        applyFilters(lyrics, sort, time, tag)
    }.stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), emptyList())

    val filteredFollowingLyrics: StateFlow<List<Lyric>> = combine(
        _followingLyrics, _sortOption, _timeRange, _selectedTag
    ) { lyrics, sort, time, tag ->
        applyFilters(lyrics, sort, time, tag)
    }.stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), emptyList())

    private var searchJob: Job? = null

    fun setSearchQuery(query: String) {
        _searchQuery.value = query
        if (query.isBlank()) {
            _searchResults.value = SearchResults()
            _isSearching.value = false
            searchJob?.cancel()
        } else {
            // Debounce search
            searchJob?.cancel()
            searchJob = viewModelScope.launch {
                delay(300)
                performSearch(query)
            }
        }
    }

    fun setSortOption(option: SortOption) { _sortOption.value = option }
    fun setTimeRange(range: TimeRange) { _timeRange.value = range }
    fun setSelectedTag(tag: String?) {
        _selectedTag.value = if (_selectedTag.value == tag) null else tag
    }

    private fun applyFilters(
        lyrics: List<Lyric>,
        sort: SortOption,
        time: TimeRange,
        tag: String?
    ): List<Lyric> {
        var result = lyrics

        // Time range filter
        val now = Clock.System.now()
        result = when (time) {
            TimeRange.TODAY -> result.filter { it.createdAt >= now - 24.hours }
            TimeRange.THIS_WEEK -> result.filter { it.createdAt >= now - 7.days }
            TimeRange.THIS_MONTH -> result.filter { it.createdAt >= now - 30.days }
            TimeRange.ALL_TIME -> result
        }

        // Tag filter
        if (tag != null) {
            result = result.filter { lyric ->
                lyric.tags?.any { it.equals(tag, ignoreCase = true) } == true
            }
        }

        // Sort
        result = when (sort) {
            SortOption.MOST_RESONATED -> result.sortedByDescending { it.reactionCount ?: 0 }
            SortOption.MOST_DISCUSSED -> result.sortedByDescending { it.commentCount ?: 0 }
            SortOption.NEWEST -> result.sortedByDescending { it.createdAt }
        }

        return result
    }

    private fun computeTrendingTags(lyrics: List<Lyric>) {
        val tagCounts = mutableMapOf<String, Int>()
        lyrics.forEach { lyric ->
            lyric.tags?.forEach { tag ->
                tagCounts[tag] = (tagCounts[tag] ?: 0) + 1
            }
        }
        _trendingTags.value = tagCounts.entries
            .sortedByDescending { it.value }
            .take(10)
            .map { it.key to it.value }
    }

    private suspend fun performSearch(query: String) {
        _isSearching.value = true
        val blocked = blockManager.blockedUserIds.value
        try {
            // Search lyrics by content, song title, artist name
            val q = query.trim()
            val matchingLyrics = supabase.postgrest.from("lyrics")
                .select {
                    filter {
                        eq("is_current", true)
                        eq("is_public", true)
                        or {
                            ilike("content", "%$q%")
                            ilike("song_title", "%$q%")
                            ilike("artist_name", "%$q%")
                        }
                    }
                    order("created_at", Order.DESCENDING)
                    limit(30)
                }
                .decodeList<Lyric>()
                .filter { it.userId !in blocked }

            // Extract distinct songs (title + artist pairs)
            val songs = matchingLyrics
                .filter { it.songTitle != null }
                .map { it.songTitle!! to it.artistName }
                .distinctBy { "${it.first.lowercase()}|${it.second?.lowercase()}" }
                .take(10)

            // Extract distinct artists
            val artists = matchingLyrics
                .mapNotNull { it.artistName }
                .distinct()
                .filter { it.lowercase().contains(q.lowercase()) }
                .take(10)

            // Search users by username
            val matchingUsers = try {
                supabase.postgrest.from("profiles")
                    .select {
                        filter {
                            ilike("username", "%$q%")
                        }
                        limit(10)
                    }
                    .decodeList<Profile>()
                    .filter { it.id !in blocked }
            } catch (_: Exception) { emptyList() }

            // Fetch profiles for lyric authors
            fetchProfilesFor(matchingLyrics)

            _searchResults.value = SearchResults(
                songs = songs,
                artists = artists,
                users = matchingUsers,
                lyrics = matchingLyrics
            )
        } catch (_: Exception) {
            _searchResults.value = SearchResults()
        } finally {
            _isSearching.value = false
        }
    }

    fun toggleReaction(lyricId: String) {
        val userId = authManager.userId ?: return
        val wasReacted = lyricId in _reactedLyricIds.value
        // Optimistic update
        _reactedLyricIds.value = if (wasReacted) _reactedLyricIds.value - lyricId else _reactedLyricIds.value + lyricId
        _reactionCountDeltas.value = _reactionCountDeltas.value + (lyricId to ((_reactionCountDeltas.value[lyricId] ?: 0) + if (wasReacted) -1 else 1))
        viewModelScope.launch {
            try {
                if (wasReacted) {
                    supabase.postgrest.from("reactions").delete { filter { eq("lyric_id", lyricId); eq("user_id", userId) } }
                } else {
                    supabase.postgrest.from("reactions").insert(ReactionInsert(lyricId, userId))
                }
            } catch (_: Exception) {
                // Revert on failure
                _reactedLyricIds.value = if (wasReacted) _reactedLyricIds.value + lyricId else _reactedLyricIds.value - lyricId
                _reactionCountDeltas.value = _reactionCountDeltas.value + (lyricId to ((_reactionCountDeltas.value[lyricId] ?: 0) + if (wasReacted) 1 else -1))
            }
        }
    }

    private suspend fun fetchUserReactions(lyricIds: List<String>) {
        val userId = authManager.userId ?: return
        if (lyricIds.isEmpty()) return
        try {
            val reactions = supabase.postgrest.from("reactions")
                .select { filter { eq("user_id", userId); isIn("lyric_id", lyricIds) } }
                .decodeList<Reaction>()
            _reactedLyricIds.value = _reactedLyricIds.value + reactions.map { it.lyricId }.toSet()
        } catch (_: Exception) { }
    }

    init {
        loadData()
    }

    fun selectTab(tab: Int) {
        _selectedTab.value = tab
    }

    fun loadData() {
        viewModelScope.launch {
            _isLoading.value = true
            try {
                fetchForYou()
                fetchFollowing()
            } finally {
                _isLoading.value = false
            }
        }
    }

    private suspend fun fetchForYou() {
        val userId = authManager.userId ?: return
        val blocked = blockManager.blockedUserIds.value
        try {
            val result = supabase.postgrest.from("lyrics")
                .select {
                    filter {
                        eq("is_current", true)
                        eq("is_public", true)
                        neq("user_id", userId)
                    }
                    order("created_at", Order.DESCENDING)
                    limit(50)
                }
                .decodeList<Lyric>()
            _forYouLyrics.value = result.filter { it.userId !in blocked }
            computeTrendingTags(_forYouLyrics.value)
            fetchProfilesFor(result)
            fetchUserReactions(_forYouLyrics.value.map { it.id })
        } catch (_: Exception) { }
    }

    private suspend fun fetchFollowing() {
        val userId = authManager.userId ?: return
        val follows = followManager.follows.value
        val userFollows = followManager.userFollows.value
        val blocked = blockManager.blockedUserIds.value

        try {
            val allLyrics = mutableListOf<Lyric>()

            // Lyrics from followed users
            val followedUserIds = userFollows.map { it.followingId }
            if (followedUserIds.isNotEmpty()) {
                val userLyrics = supabase.postgrest.from("lyrics")
                    .select {
                        filter {
                            eq("is_current", true)
                            eq("is_public", true)
                            isIn("user_id", followedUserIds)
                        }
                        order("created_at", Order.DESCENDING)
                    }
                    .decodeList<Lyric>()
                allLyrics.addAll(userLyrics)
            }

            // Lyrics matching followed artists/songs/tags
            follows.forEach { follow ->
                try {
                    val matchingLyrics = when (follow.filterType) {
                        "artist" -> supabase.postgrest.from("lyrics")
                            .select {
                                filter {
                                    eq("is_current", true)
                                    eq("is_public", true)
                                    eq("artist_name", follow.filterValue)
                                }
                                limit(10)
                            }
                            .decodeList<Lyric>()
                        "song" -> supabase.postgrest.from("lyrics")
                            .select {
                                filter {
                                    eq("is_current", true)
                                    eq("is_public", true)
                                    eq("song_title", follow.filterValue)
                                }
                                limit(10)
                            }
                            .decodeList<Lyric>()
                        else -> emptyList()
                    }
                    allLyrics.addAll(matchingLyrics)
                } catch (_: Exception) { }
            }

            _followingLyrics.value = allLyrics
                .filter { it.userId !in blocked && it.userId != userId }
                .distinctBy { it.id }
                .sortedByDescending { it.createdAt }
            fetchProfilesFor(allLyrics)
            fetchUserReactions(_followingLyrics.value.map { it.id })
        } catch (_: Exception) { }
    }

    fun getBlockManager(): BlockManager = blockManager
    fun getReporterId(): String? = authManager.userId

    private suspend fun fetchProfilesFor(lyrics: List<Lyric>) {
        val userIds = lyrics.map { it.userId }.distinct()
            .filter { it !in _lyricProfiles.value }
        if (userIds.isEmpty()) return
        try {
            val profiles = supabase.postgrest.from("profiles")
                .select { filter { isIn("id", userIds) } }
                .decodeList<Profile>()
            _lyricProfiles.value = _lyricProfiles.value + profiles.associateBy { it.id }
        } catch (_: Exception) { }
    }
}
